### PR TITLE
Raise a descriptive error if `inputs` are not inputs

### PR DESCRIPTION
**Pull request by bstriner**
Raise a descriptive error if `Model` constructor `inputs` are not inputs.

I had a typo where I swapped the inputs and outputs and the resulting error was a little weird.

> AttributeError: 'Dense' object has no attribute 'is_placeholder'

PR adds a check for `is_placeholder` just before that line so I get a message like this.

> ValueError: Input 0 (0-indexed) `<keras.layers.core.Dense object at 0x000000003C445128>` has no attribute `is_placeholder`. Is it an Input layer?

Cheers

### CODE REVIEW

File: keras/engine/topology.py [hunk 1]
```diff
@@ -1608,6 +1608,11 @@ def __init__(self, inputs, outputs, name=None):
         self._feed_input_shapes = []
         for i, layer in enumerate(self.input_layers):
             self.input_names.append(layer.name)
```
Comment (fchollet) [inline keras/engine/topology.py]: This statement should come afterwards

File: keras/engine/topology.py [hunk 2]
```diff
@@ -1608,6 +1608,11 @@ def __init__(self, inputs, outputs, name=None):
...
+            if not hasattr(layer, 'is_placeholder'):
+                raise ValueError(
+                    'Input {} (0-indexed) `{}` has no attribute `is_placeholder`. Is it an Input layer?'.format(
```
Comment (fchollet) [inline keras/engine/topology.py]: Please break up this line into several

File: keras/engine/topology.py [hunk 3]
```diff
@@ -1608,6 +1608,11 @@ def __init__(self, inputs, outputs, name=None):
         self._feed_input_shapes = []
         for i, layer in enumerate(self.input_layers):
             self.input_names.append(layer.name)
+            if not hasattr(layer, 'is_placeholder'):
+                raise ValueError(
```
Comment (fchollet) [inline keras/engine/topology.py]: The ValueError should be reported in the docstring

**Comment by fchollet** [comment]
I think we should use a simpler input message; "is_placeholder" is an implementation detail. Maybe "is not an input layer?"

**Comment by bstriner** [comment]
@fchollet I was kind of surprised it got that far when it wasn't even an input layer.

An alternative would be to add something like `if not isinstance(layer, InputLayer) ...` and make that earlier in the code. If you want all inputs to be `InputLayer` objects or something else specific that would be a helpful assertion to make.

If we don't do static typing and stick to duck typing, then I think the message still needs to mention "is_placeholder" so someone can add that attribute to their custom duck-typed whatever.

File: keras/engine/topology.py (hunk 3 above)
Comment (bstriner) [inline keras/engine/topology.py]: Any preference on the type or is ValueError a good-enough catch-all?

**Comment by fchollet** [comment]
What's the viability of a model where an input is not an `InputLayer`? Can we make it work?

If not, then we should explicitly require inputs to be `InputLayer`. If it is viable, then we should make the attribute check optional. In any case, it shouldn't come up in the error message.

**Comment by bstriner** [comment]
@fchollet not sure if there were any plans for something other than `InputLayer` but it seems like everything should be an `InputLayer` for now. I can just add an assertion that the inputs are an instance of the correct class.

**Comment by fchollet** [comment]
Thanks. We should raise a `ValueError` with a descriptive and actionable message.

**Comment by bstriner** [comment]
@fchollet It looks like from some of the earlier blocks we're raising a TypeError, which I guess makes more sense.

The code originally raised a TypeError if the inputs are not Keras tensors. Now it also raises a TypeError if the layer in the history of the input is not an `InputLayer`.

Message is ``Input layers to a `{}` must be `InputLayer` objects. Input {} (0-based) is a `{}` object``.

Added to the docstring of the class since `__init__` doesn't have its own docstring.

Cheers

**Pull request merged**